### PR TITLE
ChatMessageManager.ShowComposeSmsMessageAsync

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): #
#747 

## PR Type
What kind of change does this PR introduce?

- Feature

## What is the current behavior?
`ChatMessageManager.ShowComposeSmsMessageAsync` is not implemented

## What is the new behavior?
`ChatMessageManager.ShowComposeSmsMessageAsync` is implemented 😉 . Also required `ChatMessage` and its `Recipients` and `Body` properties.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

## Other information
Needed to introduce `NotNullList` which mimics the UWP behavior of not allowing `null` entries in the `Recipients` property. Should be useful for other places as well (for example `FileOpenPicker's` `Extensions` list if I recall correctly. Not sure about the namespace where to have it though. If some other location would be preferred, I can change it.

Internal Issue (If applicable):

### CODE REVIEW

**Comment by carldebilly** [comment]
Suggestion : having a sample testing for availability of SMS API (implemented and supported on current hardware).

File: src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.iOS.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+				throw new ArgumentNullException(nameof(message));
+			}
+
+			var window = UIApplication.SharedApplication.KeyWindow;
+			var controller = window.RootViewController;
+
+			var messageController = new MFMessageComposeViewController();
+
+			messageController.Body = message.Body;
+			messageController.Recipients = message?.Recipients?.ToArray() ?? new string[] { };
```
Comment (davidjohnoliver) [inline src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.iOS.cs]: Same question here. From the [docs](https://developer.apple.com/documentation/messageui/mfmessagecomposeviewcontroller/1614071-recipients) the Apple API seems to be more restricted:

> Each string in the array should contain the phone number of the intended recipient.

File: src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.iOS.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+			var tcs = new TaskCompletionSource<bool>();
+			messageController.Finished += (sender, e) =>
```
Comment (davidjohnoliver) [inline src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.iOS.cs]: What happens if the user cancels sending the message? What if there's an error? (No network, invalid recipient, etc.) The task should still complete.

File: src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.Android.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+				throw new ArgumentNullException(nameof(message));
+			}
+
+			var intent = CreateIntent(message.Body, message.Recipients.ToArray())
+				.SetFlags(ActivityFlags.ClearTop)
+				.SetFlags(ActivityFlags.NewTask);
+
+			ContextHelper.Current.StartActivity(intent);
+			
+			return Task.FromResult(true).AsAsyncAction();
```
Comment (davidjohnoliver) [inline src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.Android.cs]: On iOS the task completes once the message has been sent, on Android it completes synchronously as soon as the intent is launched... which is the correct behaviour? ('Correct' == matching UWP)

**Comment by MartinZikmund** [comment]
> Suggestion : having a sample testing for availability of SMS API (implemented and supported on current hardware).

@carldebilly How would you suggest implementing this? UWP docs use the API `ApiInformation.IsTypePresent("Windows.ApplicationModel.Chat.ChatMessageManager")` to check if the device is SMS capable, but I am not sure that can be utilized in Uno (apart from branching the logic in the `IsTypePresent` implementation?

File: src/Uno.Foundation/Uno.Foundation.csproj [hunk 4]
```diff
@@ -35,6 +35,10 @@
 	<ItemGroup Condition="'$(TargetFrameworkIdentifier)'=='.NETStandard'">
 		<PackageReference Include="System.Runtime.InteropServices.WindowsRuntime" Version="4.3.0" />
 	</ItemGroup>
+
+	<ItemGroup>
+	  <UpToDateCheckInput Remove="Collections\NonNullList.cs" />
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.Foundation/Uno.Foundation.csproj]: Oh! Missed that, will fix

File: src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.Android.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		private const string RecipientsSeparator = ";";
+
+		public static IAsyncAction ShowComposeSmsMessageAsync(ChatMessage message)
+		{
+			if (message == null)
+			{
+				throw new ArgumentNullException(nameof(message));
+			}
+
+			var intent = CreateIntent(message.Body, message.Recipients.ToArray())
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.Android.cs]: Actually I have tested with arbitrary strings and it did work and threw no exceptions - the "non-number" recipients appear as normal text in the _To_ field of the SMS app. Not sure if it is useful, but it is probably the same thing as in UWP

File: src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.iOS.cs (hunk 1 above)
Comment (MartinZikmund) [reply src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.iOS.cs]: Similarly on iOS it outputs non-numeric recipients as normal string in the _To_ field

File: src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.Android.cs (hunk 3 above)
Comment (MartinZikmund) [reply src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.Android.cs]: That is a good point. I will try to dig up my old Lumia to check. On desktop this just opens the deprecated SMS app, so it is not very useful and can't really be used to check.

File: src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.iOS.cs (hunk 2 above)
Comment (MartinZikmund) [reply src/Uno.UWP/ApplicationModel/Chat/ChatMessageManager.iOS.cs]: The truth is I have closely followed the implementation of SMS in Xamarin.Essentials and this was the approach chosen there. I will check with UWP and try to do the same thing

**Comment by jeromelaban** [comment]
> How would you suggest implementing this?

@MartinZikmund for now, testing with `IsTypePresent` in the sample is probably enough. The supported platforms will raise their own exceptions there, and that's fine.

**Comment by MartinZikmund** [comment]
> > How would you suggest implementing this?
> 
> @MartinZikmund for now, testing with `IsTypePresent` in the sample is probably enough. The supported platforms will raise their own exceptions there, and that's fine.

@jeromelaban I'm sorry, but I am not sure what you meant. `IsTypePresent` will always return `true` on Android and iOS, so it is a different behavior than in case of UWP where it returns false in case device does not have SMS capability. Is that ok?

**Comment by jeromelaban** [comment]
@MartinZikmund What I mean is that the behavior of the IsTypePresent is appropriate and does not need to change. 

When the running device does not provide the proper support, there will be exceptions raised when sending an SMS, which is also appropriate. Is it what happens when sending an sms on windows desktop ?

**Comment by MartinZikmund** [comment]
> @MartinZikmund What I mean is that the behavior of the IsTypePresent is appropriate and does not need to change.
> 
> When the running device does not provide the proper support, there will be exceptions raised when sending an SMS, which is also appropriate. Is it what happens when sending an sms on windows desktop ?

Surprisingly not 😊 - the API actually works on desktop as well and it just opens the deprecated Messages app and which shows "nothing". It works well on Lumia though.

**Pull request merged**